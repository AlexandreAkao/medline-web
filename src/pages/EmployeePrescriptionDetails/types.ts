export enum UpdateRequestEnum {
  'accept' = 1,
  'refuse' = 2,
}
