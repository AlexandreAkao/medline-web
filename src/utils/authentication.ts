import { isEmpty } from 'ramda';

export const getUserInfo = () => {
  const user = JSON.parse(localStorage.getItem('user') ?? '{}');

  return isEmpty(user) ? null : (user as IUser);
};

export const geToken = () => localStorage.getItem('token');
