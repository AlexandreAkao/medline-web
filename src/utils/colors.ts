export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
