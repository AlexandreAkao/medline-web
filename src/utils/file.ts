export const downloadFile = (blob: Blob, filename: string) => {
  const link = document.createElement('a');
  link.href = URL.createObjectURL(blob);
  link.download = filename;
  link.click();
};
