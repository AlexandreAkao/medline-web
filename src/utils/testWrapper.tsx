import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component: React.ReactNode) =>
  render(<MemoryRouter>{component}</MemoryRouter>);
