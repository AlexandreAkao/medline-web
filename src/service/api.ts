import axios, { AxiosError } from 'axios';
import { toast } from 'react-toastify';

import { getToken } from 'utils/authentication';

const api = axios.create({
  baseURL: (import.meta.env.VITE_BFF_BASE_URL as string) ?? '',
});

api.interceptors.request.use(
  config => {
    const authorization = getToken() ? { Authorization: `Bearer ${getToken()}` } : null;
    return {
      ...config,
      headers: {
        ...authorization,
      },
    };
  },
  error => {
    return Promise.reject(error);
  },
);

api.interceptors.response.use(
  response => {
    return response;
  },
  (error: AxiosError) => {
    if (error.response?.status === 401) {
      axios
        .post<IRefreshToken>(
          'https://securetoken.googleapis.com/v1/token',
          {
            grant_type: 'refresh_token',
            refresh_token: localStorage.getItem('refresh_token') ?? '',
          },
          {
            params: {
              key: 'AIzaSyDYolHTkT0zuI3X1MNup36eSi2gZ3INa2g',
            },
          },
        )
        .then(data => {
          const { access_token: accessToken } = data.data;

          localStorage.setItem('token', accessToken);
        });
    } else if (error.response?.status !== 404)
      if (error.response && error.response.status >= 500)
        toast.error(error.message, { autoClose: 3000, theme: 'colored' });
    return Promise.reject(error);
  },
);

export default api;
