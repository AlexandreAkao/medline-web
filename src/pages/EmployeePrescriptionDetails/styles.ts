import styled from 'styled-components';

import { ButtonContainer } from 'components/Button/styles';
import colors from 'styles/colors';

export const EmployeePrescriptionDetailsContainer = styled.div`
  background-color: ${colors.white.lighter};
  height: 100%;
  background-repeat: no-repeat;
  background-attachment: fixed;
  background-position-x: 50%;
  overflow: auto;
`;

export const EmployeePrescriptionDetailsMain = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 2rem auto;
  width: 80%;
`;

export const EmployeePrescriptionDetailsTitle = styled.h1`
  font-size: 2.125rem;
  font-weight: 500;
  margin-bottom: 2rem;
`;

export const EmployeePrescriptionDetailsInformation = styled.main`
  border-radius: 1rem;
  background-color: ${colors.white.normal};
  width: 100%;
  box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.2);
  overflow: hidden;
`;

export const EmployeePrescriptionDetailsInformationData = styled.div`
  background-color: ${colors.primaryLighter};
  padding: 1rem;
  display: flex;
  align-items: center;
  justify-content: space-around;

  span {
    font-weight: 600;
  }
`;

export const EmployeePrescriptionDetailsInformationDescription = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;

  h1 {
    margin-top: 1rem;
    font-weight: 500;
  }

  span {
    border-top: 1px solid ${colors.black.soft};
    width: 100%;
    text-align: center;
    font-size: 1.5rem;
    margin-top: 1rem;
    padding: 1rem 0 20rem 0;
  }
`;

export const EmployeePrescriptionDetailsVizualizer = styled.button`
  margin-top: 2rem;

  text-decoration: underline;
  color: ${colors.blue.normal};
  font-size: 1.125rem;
  :hover {
    color: ${colors.blue.light};
  }
`;

export const EmployeePrescriptionDetailsButtons = styled.div`
  margin-top: 2rem;
  display: flex;
  justify-content: space-between;
  width: 100%;

  ${ButtonContainer} {
    padding: 1.25rem 2rem;
  }

  div {
    display: flex;
    justify-content: center;
    gap: 2rem;
  }
`;
