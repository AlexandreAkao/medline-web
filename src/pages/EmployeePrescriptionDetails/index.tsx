import axios from 'axios';
import { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';

import Button from 'components/Button';
import MedlineHeader from 'components/Header/MedlineHeader';
import { EmployeePatienteCareTitle } from 'pages/EmployeePatienteCare/styles';
import {
  EmployeePrescriptionDetailsButtons,
  EmployeePrescriptionDetailsContainer,
  EmployeePrescriptionDetailsInformation,
  EmployeePrescriptionDetailsInformationData,
  EmployeePrescriptionDetailsInformationDescription,
  EmployeePrescriptionDetailsMain,
  EmployeePrescriptionDetailsVizualizer,
} from 'pages/EmployeePrescriptionDetails/styles';
import api from 'service/api';

function EmployeePrescriptionDetails() {
  const navigate = useNavigate();

  const downloadFile = async () => {
    const file = await api.get('file/3', {
      responseType: 'blob',
      headers: {
        Authorization:
          'Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6IjFhZWY1NjlmNTI0MTRlOWY0YTcxMDRiNmQwNzFmMDY2ZGZlZWQ2NzciLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJodHRwczovL3NlY3VyZXRva2VuLmdvb2dsZS5jb20vbWVkbGluZS1hdXRoIiwiYXVkIjoibWVkbGluZS1hdXRoIiwiYXV0aF90aW1lIjoxNjU1NDE1MDg5LCJ1c2VyX2lkIjoiRUtMQThuTGhtZlpPckhYV2VmdkRvemVHMFF5MiIsInN1YiI6IkVLTEE4bkxobWZaT3JIWFdlZnZEb3plRzBReTIiLCJpYXQiOjE2NTU0MTUwODksImV4cCI6MTY1NTQxODY4OSwiZW1haWwiOiJmdW5jaW9uYXJpb0BtZWRsaW5lLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwiZmlyZWJhc2UiOnsiaWRlbnRpdGllcyI6eyJlbWFpbCI6WyJmdW5jaW9uYXJpb0BtZWRsaW5lLmNvbSJdfSwic2lnbl9pbl9wcm92aWRlciI6InBhc3N3b3JkIn19.ikdKPc6zmAP7dBd10kQTGNzpEQjQVooJra-9EtK-5lxGO_cFkkPCezMfW5cEggDvjfbgbgpR4cyViAp98xMsjhCX3qYjq86bieDzY043FwN5RT5Lt1UiMysFaN6FjXfR9rZLKPeTChBdiY-5xA11MC42CJfvHzajoyy4PUbEsEvrZxzD4w9h3G1Z06A6xcEp9bnBPjpmi6ROlS80H798Jy669I10Oy1mYaUmIh4f1o-sJOTUq7DRZKE9pLzhy8v3J1QavFZOQ0kCnbxWZH3Y9CA0mf-1c7IvYS1DVEXgJPangJchv0s0JVWgmna4_IeOIxQFwzna89gG7DKNPB31AA',
      },
    });
    const filename = file.headers['content-disposition'].split('filename=')[1];
    const url = window.URL.createObjectURL(file.data);
    const link = document.createElement('a');
    link.href = url;
    link.setAttribute('download', filename);
    document.body.appendChild(link);
    link.click();
  };

  const handleGoBack = () => {
    navigate(-1);
  };

  return (
    <EmployeePrescriptionDetailsContainer>
      <MedlineHeader />
      <EmployeePrescriptionDetailsMain>
        <EmployeePatienteCareTitle>Atestados e Receitas</EmployeePatienteCareTitle>

        <EmployeePrescriptionDetailsInformation>
          <EmployeePrescriptionDetailsInformationData>
            <span>Gilson Bezerra</span>
            <span>000.000.000-00</span>
            <span>Solicitação: Atestado</span>
            <span>Data de envio: 18/03/22</span>
          </EmployeePrescriptionDetailsInformationData>

          <EmployeePrescriptionDetailsInformationDescription>
            <h1>Descrição</h1>
            <span>Desejo um atestado médico, pois testei positivo para covid-19</span>
          </EmployeePrescriptionDetailsInformationDescription>
        </EmployeePrescriptionDetailsInformation>

        <EmployeePrescriptionDetailsVizualizer onClick={downloadFile}>
          Baixar documento
        </EmployeePrescriptionDetailsVizualizer>

        <EmployeePrescriptionDetailsButtons>
          <Button size="large" onClick={handleGoBack}>
            Voltar
          </Button>
          <div>
            <Button size="large">Negado</Button>
            <Button size="large">Enviar</Button>
          </div>
        </EmployeePrescriptionDetailsButtons>
      </EmployeePrescriptionDetailsMain>
    </EmployeePrescriptionDetailsContainer>
  );
}

export default EmployeePrescriptionDetails;
