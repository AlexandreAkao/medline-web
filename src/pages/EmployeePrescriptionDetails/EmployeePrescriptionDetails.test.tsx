import { screen } from '@testing-library/react';
import { vi } from 'vitest';

import { renderWithProvider } from 'utils/testWrapper';
import EmployeePrescriptionDetails from 'pages/EmployeePrescriptionDetails';
import { requestMock } from '__tests__/mocks/request.mock';
import api from 'service/api';

vi.mock('react-router-dom', async () => {
  const reactRouterDom = await vi.importActual('react-router-dom');

  return {
    ...(reactRouterDom as Record<string, unknown>),
    useParams: () => ({ id: '1' }),
  };
});

describe('EmployeePrescriptionDetails', () => {
  beforeEach(() => {
    vi.spyOn(api, 'get').mockImplementation(() =>
      Promise.resolve({
        data: requestMock,
      }),
    );
  });

  const renderComponent = () => {
    return renderWithProvider(<EmployeePrescriptionDetails />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    renderComponent();
    const titleElement = screen.getByText('Atestados e Receitas');

    expect(titleElement).toBeInTheDocument();
  });
});
