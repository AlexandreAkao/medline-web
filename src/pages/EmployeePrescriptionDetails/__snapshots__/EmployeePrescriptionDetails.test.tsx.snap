// Vitest Snapshot v1

exports[`EmployeePrescriptionDetails > should match with snapshot 1`] = `
<div>
  <div
    class="sc-hAZoDl dERzXv"
  >
    <header
      class="sc-dkzDqf WepAw"
      data-testid="test-navigation"
    >
      <img
        alt="Medline logo"
        class="sc-hKMtZM hOePDa"
        src="/src/assets/images/medline-logo.svg"
      />
      <nav
        class="sc-eCYdqJ dDvGMN"
        data-testid="test-navigation-items"
      >
        <div
          class="sc-gsnTZi qZpGk"
        >
          <button
            class="sc-bczRLJ eeLoKZ"
          >
            Cadastro
          </button>
        </div>
        <div
          class="sc-gsnTZi qZpGk"
        >
          <a
            href="/login"
          >
            Entrar
          </a>
        </div>
      </nav>
    </header>
    <div
      class="sc-fnykZs bAYsYk"
    >
      <h1
        class="sc-ksZaOG fvoegn"
      >
        Atestados e Receitas
      </h1>
      <main
        class="sc-bjUoiL cgFIRp"
      >
        <div
          class="sc-idiyUo bsnWRQ"
        >
          <span>
            Gilson Bezerra
          </span>
          <span>
            000.000.000-00
          </span>
          <span>
            Solicitação: Atestado
          </span>
          <span>
            Data de envio: 18/03/22
          </span>
        </div>
        <div
          class="sc-dIouRR TPyNd"
        >
          <h1>
            Descrição
          </h1>
          <span>
            Desejo um atestado médico, pois testei positivo para covid-19
          </span>
        </div>
      </main>
      <button
        class="sc-hHLeRK hgoXaG"
      >
        Baixar documento
      </button>
      <div
        class="sc-dmRaPn djFJY"
      >
        <button
          class="sc-bczRLJ fpIqOv"
        >
          Voltar
        </button>
        <div>
          <button
            class="sc-bczRLJ fpIqOv"
          >
            Negado
          </button>
          <button
            class="sc-bczRLJ fpIqOv"
          >
            Enviar
          </button>
        </div>
      </div>
    </div>
  </div>
  ,
</div>
`;
